src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -697,3 +721,84 @@ impl rwh_06::HasDisplayHandle for OwnedDisplayHandle {
         Ok(unsafe { rwh_06::DisplayHandle::borrow_raw(raw.into()) })
     }
 }
+
+#[derive(Debug)]
+struct PumpEventNotifier {
+    /// Whether we're in winit or not.
+    control: Arc<(Mutex<PumpEventNotifierAction>, Condvar)>,
+    /// Waker handle for the working thread.
+    worker_waker: Option<OwnedFd>,
+    /// Thread handle.
+    handle: Option<JoinHandle<()>>,
+}
+
+impl Drop for PumpEventNotifier {
+    fn drop(&mut self) {
+        // Wake-up the thread.
+        if let Some(worker_waker) = self.worker_waker.as_ref() {
+            let _ = rustix::io::write(worker_waker.as_fd(), &[0u8]);
+        }
+        *self.control.0.lock().unwrap() = PumpEventNotifierAction::Monitor;
+        self.control.1.notify_one();
+
+        if let Some(handle) = self.handle.take() {
+            let _ = handle.join();
+        }
+    }
+}
+
+impl PumpEventNotifier {
+    fn spawn(connection: Connection, awakener: Ping) -> Self {
+        // Start from the waiting state.
+        let control = Arc::new((Mutex::new(PumpEventNotifierAction::Pause), Condvar::new()));
+        let control_thread = Arc::clone(&control);
+
+        let (read, write) = match pipe::pipe_with(PipeFlags::CLOEXEC | PipeFlags::NONBLOCK) {
+            Ok((read, write)) => (read, write),
+            Err(_) => return Self { control, handle: None, worker_waker: None },
+        };
+
+        let handle =
+            std::thread::Builder::new().name(String::from("pump_events mon")).spawn(move || {
+                let (lock, cvar) = &*control_thread;
+                'outer: loop {
+                    let mut wait = lock.lock().unwrap();
+                    while *wait == PumpEventNotifierAction::Pause {
+                        wait = cvar.wait(wait).unwrap();
+                    }
+                    // Wake-up the main loop and put this one back to sleep.
+                    *wait = PumpEventNotifierAction::Pause;
+                    drop(wait);
+
+                    while let Some(read_guard) = connection.prepare_read() {
```

Comment:
Does this also ensure that all wayland-client event queues are empty?

Reply:
it just calls the `prepare_read_queue` on the `wl_event_queue` used by `wayland-client`, so the logic is exactly as you'd use the C calls directly.

it just, the issue was with `read` before since it had `dispatch` unconditionally.

The logic right now is that if you have something in the queue (you read it before with `read_events`) then it'll go to notifying the loop and going back to sleep.

Reply:
But generally, if the queue was not empty, we'll just notify the event loop without doing anything extra. The `None` is returned when you have something in the `queue` (as you'd get `-1`).

Reply:
>it just calls the prepare_read_queue on the wl_event_queue used by wayland-client, so the logic is exactly as you'd use the C calls directly.

In C there is only one queue layer. It wayland-client there are two queue layers and you have to make sure that both are empty before going to sleep.

Reply:
Do you mean the internal `libwayland-client` one? Or do you mean the `wl_event_queue` in `wayland-rs` and whatever other abstraction you have?

In general, once we re-enter winit we dispatch everything we're read on the pump events thread, there could be still not read events though (on socket side), but as for how this thread will go, I don't think it's an issue. Things can not move into `wayland-rs` rust side queue until you call the `dispatch_inner_queue` which only the main thread with `pump_events` calls.

Reply:
The queue this thread touches is only the `wl_event_queue` we've created with the `C` API, it doesn't interact with any other queue.

The `pump_events` thread, is the one pulling from `wl_event_queue` into the `Queue`, but it dispatches everything straight away.

The only difference on this thread compared to C, is that we don't use default queue, but I think it shouldn't matter?

Reply:
Though, I'm not sure how the wiring that `wayland-rs` does interacts when other thread pulls everything, I'd assume that it'll end up in `wl_event_queue`, and only once we call the `dispatch_inner_queue` it'll advance further.

Reply:
>The pump_events thread (where EventLoop is), is the one pulling from wl_event_queue into the Queue, but it dispatches everything straight away.

This still relies on there only being one thread and one Queue. Otherwise any other thread could call dispatch_inner which would read events into any other Queue. It is fragile in a way that the C API is not.

Reply:
All `Queue` share the same `wl_event_queue`, if you pass anything `wayland-rs` to ffi and create from it, it'll create a new `wl_event_queue`, so they are all isolated.

So effectively, if you use `wayland-rs` with winit and create from its `ffi` object, you create a new `wl_event_queue` and so can not mess up the queue in winit. The only issue could happen _if winit will share the rust objects directly, which will likely never happen.

Reply:
>The only issue could happen _if winit will share the rust objects directly, which will likely never happen.

It could also happen if there is more than one Queue. I.e. the wayland-client API cannot be used if an application uses multiple Queues.

Reply:
Are you talking about general issue or the issue with the concret implementation right now, since right now I don't see how there could be an issue given that winit is isolated from any other code.

Like any user will have different C queue, even if you test with `pump_events` example it also has 2 queues, the one in `softbuffer` and the one in `winit` and there's no issue with such setup.

Reply:
>Are you talking about general issue or the issue with the concret implementation right now, since right now I don't see how there could be an issue given that winit is isolated from any other code.

I don't know what winit does. But anything that relies on there never being more than 1 event queue seems questionable.

Reply:
But right now we already have more than one queue in tests, but queues isolated from each other... The issue is only if I within winit's code start adding extra stuff within itself and I somehow do the polling the way it was originally done on this PR.

I don't see how adding extra C queue outside of winit will mess up winit. It'll basically read into `wl_event_queue` (C object) winit has and then we'll start getting `None` in `prepare_read` (the way it worked with blocking poll).

Reply:
I'm not talking about wl_event_queue. I'm talking about wayland-client queues.

Reply:
But they are backed by `wl_event_queue`, and once you create new `Connection` from `ptr` it's backed by new `wl_event_queue`. It's true that multiple `Queue` refer to the same `wl_event_queue`, but they are encapsulated by `Connection` object. So you can not mess with winit `Queue` outside of winit, since you are all separated by `Connection`.

Reply:
I'm aware.

Reply:
yeah, I'm then not sure how winit state could be messed up by some other thread, since only winit can call into `dispatch_inner_queue` to dispatch its `wl_event_queue`.

Reply:
I'm saying that it is fragile in a way that the C API is not. The C API supports any number of threads using any number of queues. The wayland-client API does not. If anyone at any point decides to add more queues to to winit, the polling will silently break.

Reply:
Yeah, that part I understand, so we agree on that point. I'm just saying that in the current state of things there's no _issue_. It's true that someone will mess up with queues on my behalf and not notice that example broke (if you add a queue the thing that will broke is the `pump_events` example straight away).